### PR TITLE
[SYCLomatic] Support migration of 6 cub::Block{RadixSort, Exchange} APIs

### PR DESCRIPTION
**Pull request by yihwang-nv**

### CODE REVIEW

File: clang/lib/DPCT/CUBAPIMigration.cpp [hunk 1]
```diff
@@ -172,11 +178,133 @@ void CubMemberCallRule::registerMatcher(ast_matchers::MatchFinder &MF) {
       this);
 }
 
+static std::pair<const VarDecl *, TypeLoc>
+getTempstorageVarAndValueTypeLoc(const CXXMemberCallExpr *MC) {
+  Expr *Obj = MC->getImplicitObjectArgument();
+  const VarDecl *TempStorage = nullptr;
+
+  auto FindTempStorageVarInCtor = [&](const Expr *E) -> const VarDecl * {
```
Comment (tangjj11) [inline clang/lib/DPCT/CUBAPIMigration.cpp]:
```suggestion
  auto FindTempStorageVarInCtor = [](const Expr *E) -> const VarDecl * {
```

File: clang/lib/DPCT/CUBAPIMigration.cpp [hunk 2]
```diff
@@ -172,11 +178,133 @@ void CubMemberCallRule::registerMatcher(ast_matchers::MatchFinder &MF) {
...
+          if (VD->hasAttr<CUDASharedAttr>() && isCubVar(VD)) {
+            return VD;
+          }
+        }
+      }
+    }
+    return nullptr;
+  };
+
+  auto HandleTypeLoc = [&](TypeLoc Loc) -> TypeLoc {
```
Comment (tangjj11) [inline clang/lib/DPCT/CUBAPIMigration.cpp]:
```suggestion
  auto HandleTypeLoc = [](TypeLoc Loc) -> TypeLoc {
```

File: clang/lib/DPCT/CUBAPIMigration.cpp [hunk 3]
```diff
@@ -172,11 +178,133 @@ void CubMemberCallRule::registerMatcher(ast_matchers::MatchFinder &MF) {
...
+      default:
+        return Loc;
+      }
+    }
+  };
+
+  TypeLoc DataTypeLoc;
+  if (const auto *MTE = dyn_cast<MaterializeTemporaryExpr>(Obj)) {
+    if (auto *TOE = dyn_cast<CXXTemporaryObjectExpr>(MTE->getSubExpr())) {
+      DataTypeLoc = HandleTypeLoc(TOE->getTypeSourceInfo()->getTypeLoc());
```
Comment (tangjj11) [inline clang/lib/DPCT/CUBAPIMigration.cpp]: The name `HandleTypeLoc` is not clear, should be `getxxx`, like: `GetTargetTypeLoc`?

File: clang/lib/DPCT/CUBAPIMigration.cpp [hunk 4]
```diff
@@ -172,11 +178,133 @@ void CubMemberCallRule::registerMatcher(ast_matchers::MatchFinder &MF) {
...
+    } else if (auto *FC = dyn_cast<CXXFunctionalCastExpr>(MTE->getSubExpr())) {
+      DataTypeLoc = HandleTypeLoc(FC->getTypeInfoAsWritten()->getTypeLoc());
+    }
+    TempStorage = FindTempStorageVarInCtor(MTE->getSubExpr()->IgnoreCasts());
+  } else if (const auto *DRE = dyn_cast<DeclRefExpr>(Obj)) {
+    if (const auto *VD = dyn_cast<VarDecl>(DRE->getDecl())) {
+      DataTypeLoc = HandleTypeLoc(VD->getTypeSourceInfo()->getTypeLoc());
+      if (isCubCollectiveRecordType(VD->getType()) && VD->hasInit()) {
+        // emplaceTransformation(new ReplaceVarDecl(VD, ""));
```
Comment (tangjj11) [inline clang/lib/DPCT/CUBAPIMigration.cpp]: Why add this line?

**Review by tangjj11 — APPROVED** [review]

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp [hunk 5]
```diff
@@ -278,45 +278,58 @@ template <typename T, int VALUES_PER_THREAD> class exchange {
...
-      keys[i] = buffer[offset];
-    }
+    blocked_offset get_blocked_offset;
+    striped_offset get_striped_offset;
+    helper_exchange(item, keys, keys, get_striped_offset, get_blocked_offset);
   }
 
   /// Rearrange elements from blocked order to striped order
   template <typename Item>
   __dpct_inline__ void blocked_to_striped(Item item,
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/group_utils.hpp:354]: it's breaking the rule "backward compatibility"

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp [hunk 6]
```diff
@@ -278,45 +278,58 @@ template <typename T, int VALUES_PER_THREAD> class exchange {
...
     striped_offset get_striped_offset;
     blocked_offset get_blocked_offset;
-    helper_exchange(item, keys, get_blocked_offset, get_striped_offset);
+    helper_exchange(item, input, output, get_blocked_offset,
+                    get_striped_offset);
   }
 
   /// Rearrange elements from striped order to blocked order
   template <typename Item>
   __dpct_inline__ void striped_to_blocked(Item item,
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/group_utils.hpp:380]: it breaks the rule:  keep backward compatibility.

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp [hunk 7]
```diff
@@ -426,6 +462,21 @@ class radix_sort {
     helper_sort(item, keys, begin_bit, end_bit, true);
   }
 
+  template <typename Item>
+  __dpct_inline__ void sort(const Item &item, T (&keys)[VALUES_PER_THREAD],
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/group_utils.hpp]: Could u add comment for the new added APIs?

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp [hunk 8]
```diff
@@ -278,45 +278,58 @@ template <typename T, int VALUES_PER_THREAD> class exchange {
...
-    }
+    striped_offset get_striped_offset;
+    blocked_offset get_blocked_offset;
+    helper_exchange(item, keys, keys, get_blocked_offset, get_striped_offset);
+  }
 
-    item.barrier(sycl::access::fence_space::local_space);
+  /// Rearrange elements from striped order to blocked order
+  template <typename Item>
+  __dpct_inline__ void striped_to_blocked(Item item,
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/group_utils.hpp:330]: Could u add comments what is striped and what is blocked?

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp [hunk 9]
```diff
@@ -278,45 +278,58 @@ template <typename T, int VALUES_PER_THREAD> class exchange {
 
   template <typename Item, typename offsetFunctorTypeFW,
             typename offsetFunctorTypeRV>
-  __dpct_inline__ void helper_exchange(Item item, T (&keys)[VALUES_PER_THREAD],
-                                       offsetFunctorTypeFW &offset_functor_fw,
-                                       offsetFunctorTypeRV &offset_functor_rv) {
+  [[deprecated("This is a internal function, use blocked_to_striped, "
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/group_utils.hpp]: is the deprecated msg necessary?

Comment (zhimingwang36) [reply clang/runtime/dpct-rt/include/dpct/group_utils.hpp]: the message should be:  Please use ... instead.

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp [hunk 10]
```diff
@@ -279,45 +279,58 @@ template <typename T, int VALUES_PER_THREAD> class exchange {
...
-    for (size_t i = 0; i < VALUES_PER_THREAD; i++) {
-      size_t offset = offset_functor_rv(item, i);
-      keys[i] = buffer[offset];
-    }
+    blocked_offset get_blocked_offset;
+    striped_offset get_striped_offset;
+    helper_exchange(item, keys, keys, get_striped_offset, get_blocked_offset);
   }
 
   /// Rearrange elements from blocked order to striped order
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/group_utils.hpp]: Please add comment on what is blocked order and what is striped

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp [hunk 11]
```diff
@@ -413,24 +452,88 @@ class radix_sort {
...
+/// Implements radix sort to sort integer data elements assigned to all threads
+/// in the group.
+///
+/// \tparam T type of the data elements exchanges
+/// \tparam VALUES_PER_THREAD number of data elements assigned to a thread
+/// \tparam DECENDING boolean value indicating if data elements are sorted in
+/// decending order.
+template <typename T, int VALUES_PER_THREAD, bool DESCENDING = false>
+class [[deprecated(
+    "This class deprecated, please use group_radix_sort instead")]] radix_sort {
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/group_utils.hpp]: use the the same style statement,  Please use ... instead.

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp [hunk 12]
```diff
@@ -358,9 +394,13 @@ template <typename T, int VALUES_PER_THREAD> class exchange {
 /// \tparam VALUES_PER_THREAD number of data elements assigned to a thread
 /// \tparam DECENDING boolean value indicating if data elements are sorted in
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/group_utils.hpp]: this param is not existed any more.

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp [hunk 13]
```diff
@@ -339,10 +352,33 @@ template <typename T, int VALUES_PER_THREAD> class exchange {
...
+
+    T *buffer = reinterpret_cast<T *>(_local_memory);
+
+#pragma unroll
+    for (size_t i = 0; i < VALUES_PER_THREAD; i++) {
+      size_t offset = offset_functor_fw(item, i);
+      buffer[offset] = input[i];
+    }
+
+    item.barrier(sycl::access::fence_space::local_space);
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/group_utils.hpp]: please use group barrier.

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp [hunk 14]
```diff
@@ -279,168 +279,394 @@ template <typename T, int VALUES_PER_THREAD> class exchange {
...
+  helper_exchange(Item item, T (&keys)[ElementsPerWorkItem],
+                  offsetFunctorTypeFW &offset_functor_fw,
+                  offsetFunctorTypeRV &offset_functor_rv) {
+    helper_exchange(item, keys, keys, offset_functor_fw, offset_functor_rv);
+  }
 
-    item.barrier(sycl::access::fence_space::local_space);
+  /// Inplace rearrange elements from blocked order to striped order.
+  ///
+  /// Suppose ElementsPerWorkItem is 4 and the blocked input across the
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/group_utils.hpp]:
```suggestion
  ///  Suppose 512(input) integer data items partitioned across 128 work items, where each work item owns 4 (ElementsPerWorkItem) data items and the blocked input across the
```

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp [hunk 15]
```diff
@@ -279,168 +279,394 @@ template <typename T, int VALUES_PER_THREAD> class exchange {
...
+  /// work-group is:
+  ///   {[0, 1, 2, 3], [4, 4, 6, 7], ..., [508, 509, 510, 511]}.
+  /// The corresponding output of each work-item will be:
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/group_utils.hpp]:
```suggestion
  /// The striped order output is: 
```

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp [hunk 16]
```diff
@@ -279,168 +279,394 @@ template <typename T, int VALUES_PER_THREAD> class exchange {
...
+                  offsetFunctorTypeRV &offset_functor_rv) {
+    helper_exchange(item, keys, keys, offset_functor_fw, offset_functor_rv);
+  }
 
-    item.barrier(sycl::access::fence_space::local_space);
+  /// Inplace rearrange elements from blocked order to striped order.
+  ///
+  /// Suppose ElementsPerWorkItem is 4 and the blocked input across the
+  /// work-group is:
+  ///   {[0, 1, 2, 3], [4, 4, 6, 7], ..., [508, 509, 510, 511]}.
```
Comment (zhimingwang36) [inline clang/runtime/dpct-rt/include/dpct/group_utils.hpp]:
```suggestion
  ///   {[0, 1, 2, 3], [4, 5, 6, 7], ..., [508, 509, 510, 511]}.
```

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp (hunk 14 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/group_utils.hpp]: done.

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp (hunk 16 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/group_utils.hpp]: done.

File: clang/runtime/dpct-rt/include/dpct/group_utils.hpp (hunk 15 above)
Comment (yihwang-nv) [reply clang/runtime/dpct-rt/include/dpct/group_utils.hpp]: done.

**Review by zhimingwang36 — APPROVED** [review]

**Pull request merged**